pkg/controller/workloadspread/update_pod_deletion_cost.go

```diff
@@ -0,0 +1,208 @@
+/*
+Copyright 2021 The Kruise Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package workloadspread
+
+import (
+	"context"
+	"sort"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog"
+	kubecontroller "k8s.io/kubernetes/pkg/controller"
+
+	appsv1alpha1 "github.com/openkruise/kruise/apis/apps/v1alpha1"
+	wsutil "github.com/openkruise/kruise/pkg/util/workloadspread"
+)
+
+func (r *ReconcileWorkloadSpread) updateDeletionCost(ws *appsv1alpha1.WorkloadSpread,
+	podMap map[string][]*corev1.Pod,
+	orphanPods []*corev1.Pod,
+	workloadReplicas int32) error {
+	// update Pod's deletion-cost annotation in each subset
+	for _, subset := range ws.Spec.Subsets {
+		if err := r.syncSubsetPodDeletionCost(ws, &subset, podMap[subset.Name], workloadReplicas); err != nil {
+			return err
+		}
+	}
+	// update orphanPods, deletion-cost = -100
+	return r.updateDeletionCostForSubsetPods(ws, "orphan", orphanPods, wsutil.PodDeletionCostNegative)
+}
+
+// syncSubsetPodDeletionCost calculates the deletion-cost for the Pods belong to subset and update deletion-cost annotation.
+// We have three types for subset's Pod deletion-cost
+// 1. the number of active Pods in this subset <= maxReplicas, deletion-cost = 100. indicating the priority of Pods
```

Comment:
if all or multiple subsets has nil maxReplicas,  deletion cost should be higher for the subset with larger index in the subsets array, e.g. one deploy pods first into ecs and if resources is not enough, deploy pods into eci. MaxReplicas can be nil for both subsets